### ISSUE TITLE
Unify Tensor and PIL transforms

### ISSUE THREAD

**Issue by fmassa** [pytorch/vision#2292]
Now that most of the tensor transforms have been implemented https://github.com/pytorch/vision/issues/1375, it is time to make unify the implementations for PIL and Tensor so that `torchvision.transforms` and `torchvision.transforms.functional` works seamlessly between both datatypes.

As such, we would like that the following work with the same interface:

```python
# functional interface
torchvision.transforms.functional.hflip(pil_image)
torchvision.transforms.functional.hflip(tensor_image)

# class interface
transform = torchvision.transforms.RandomHorizontalFlip(0.3)
transform(pil_image)
transform(tensor_image)

# torchscript support for tensor_image
script_transform = torch.jit.script(transform)
script_transform(tensor_image)
```

Example PRs adding support for `hflip` and `vflip`: https://github.com/pytorch/vision/pull/2282 https://github.com/pytorch/vision/pull/2283

We would like to add support for torchscript to the `torchvision.transforms` interface as well, which for simplicity might require a few changes. Most notably we would need the transforms to inherit from `nn.Module`.

Here is a list of transforms that can be readily converted:
- [X] adjust_brightness
- [X] adjust_contrast
- [X] adjust_saturation
- [X] crop
- [X] center_crop
- [X] five_crop
- [X] ten_crop
- [X] pad (#2373 #2345)
- [x] resize (#2394)
- [x] resized_crop (#2396)
- [x] perspective (#2558, #2561)
- [x] adjust_hue (#2566)
- [x] adjust_gamma (#2459)
- [x] rotate (#2495 #2496 #2553)
- [x] affine (#2444, #2468, #2478, #2553)
- [x] to_grayscale

**Comment by clemkoa** [comment]
Hi! I would be glad to work on a PR for this issue. I can start with the color jitter if that's okay with you

**Comment by vfdev-5** [comment]
@fmassa I can work on some of these listed transformation if it is OK.

**Comment by dimimal** [comment]
In the documentation of pytorch it is mentioned that functional module works with both tensors and pil objects. This is about torch version 1.5.1 and torchvision 0.6.1. However, when I try to use them with torch.tensor it throws an error that the object is not PIL. I checked the source code and the snippet shown in the docs is missing. I use the functional.hflip() function of the docs. I thought that the unification has been completed, hasn't it?

**Comment by fmassa** [comment]
@dimimal the unification is not yet complete

Can you point out where in the documentation you found the reference that all functions work with both tensors and PIL Images?

**Comment by fmassa** [comment]
@vfdev-5 sure go for it!

**Comment by dimimal** [comment]
@fmassa  check this [https://pytorch.org/docs/stable/_modules/torchvision/transforms/functional.html#hflip](url)

**Comment by fmassa** [comment]
@dimimal `hflip` supports both tensors and PIL Images since https://github.com/pytorch/vision/pull/2282 , but only on master and only a handful of functions support it for now

**Comment by dimimal** [comment]
@fmassa I wasn't using the master and the docs confused me with the noted version. Thank you. Is there anything I can do to help in contribution?

**Comment by fmassa** [comment]
@dimimal sure, contributions are more than welcome! There are two steps here: finishing up https://github.com/pytorch/vision/issues/1375 , and then implementing the glue code, similarly to the example PRs #2282 #2283, or more recently #2345 

Note that one of the biggest difficulties will be making the implementation work with torchscript, as not everything that we use is natively supported.

**Comment by vfdev-5** [comment]
Complete list for ops to support:
- [x] adjust_brightness
- [x] adjust_contrast
- [x] adjust_saturation
- [x] crop
- [x] center_crop
- [x] five_crop
- [x] ten_crop
- [ ]  
- [ ] 
- [ ] 
- [ ] 
- [ ]

**Comment by vfdev-5** [comment]
@fmassa if ops should also work on tensors on non-cpu devices, a supplementary check of certain geometrical ops will be needed. Should we setup some CI tests for that ?

**Comment by fmassa** [comment]
> Should we setup some CI tests for that ?

Yes, please!

**Comment by vfdev-5** [comment]
@nairbv would you be interested in working on unifying `adjust_hue` ?

**Comment by fmassa** [comment]
@vfdev-5 go for it!

**Comment by vfdev-5** [comment]
According to a discussion with @fmassa , we also have to make sure that all transformations work with a batch of images.

**Comment by vfdev-5** [comment]
The issue can be closed as all mentioned transformations now support tensor and pil image as input.

**Comment by voldemortX** [comment]
Hi! Very exciting codes! Wondering whether future commits will support fill color other than 0 for tensor affine transformations? It is kind of important for tasks like semantic segmentation when transforming labels. @vfdev-5

**Comment by vfdev-5** [comment]
@voldemortX thanks for the feedback ! 

> Wondering whether future commits will support fill color other than 0 for tensor affine transformations? It is kind of important for tasks like semantic segmentation when transforming labels.

Could you please open an issue for that such we could track this feature request and see what could be possible to do. 
Currently, the blocker is with pytorch [`grid_sample`](https://pytorch.org/docs/stable/nn.functional.html#torch.nn.functional.grid_sample) which has 
> padding_mode="zeros": use 0 for out-of-bound grid locations

**Comment by voldemortX** [comment]
> @voldemortX thanks for the feedback !
> 
> > Wondering whether future commits will support fill color other than 0 for tensor affine transformations? It is kind of important for tasks like semantic segmentation when transforming labels.
> 
> Could you please open an issue for that such we could track this feature request and see what could be possible to do.
> Currently, the blocker is with pytorch [`grid_sample`](https://pytorch.org/docs/stable/nn.functional.html#torch.nn.functional.grid_sample) which has
> 
> > padding_mode="zeros": use 0 for out-of-bound grid locations

New issue opened #2887
